test/brod_demo_group_subscriber_koc.erl

```diff
@@ -0,0 +1,138 @@
+
+-module(brod_demo_group_subscriber_koc).
+
+-behaviour(brod_group_subscriber).
+
+-export([ bootstrap/0
+        , bootstrap/1
+        ]).
+
+%% behabviour callbacks
+-export([ init/2
+        , handle_message/4
+        ]).
+
+
+-include_lib("brod/include/brod.hrl").
+
+-define(PRODUCE_DELAY_SECONDS, 5).
+
+%% @doc This function bootstraps everything to demo of group subscriber.
+%% Prerequisites:
+%%   - bootstrap docker host at {"localhost", 9092}
+%%   - kafka topic named <<"brod-group-subscriber-demo-koc">>
+%%     having two or more partitions.
+%% Processes to spawn:
+%%   - A brod client
+%%   - A producer which produces sequence numbers to each partition
+%%   - X group subscribers, X is the number of partitions
+%%
+%% * consumed sequence numbers are printed to console
+%% * consumed offsets are commited to kafka (using v2 commit requests)
+%% @end
+-spec bootstrap() -> ok.
+bootstrap() ->
+  bootstrap(?PRODUCE_DELAY_SECONDS).
+
+bootstrap(DelaySeconds) ->
+  ClientId = ?MODULE,
+  BootstrapHosts = [{"localhost", 9092}],
+  ClientConfig = [],
+  Topic = <<"brod-demo-group-subscriber-koc">>,
+  {ok, _ClientPid} =
+    brod:start_link_client(BootstrapHosts, ClientId, ClientConfig),
+  ok = brod:start_producer(ClientId, Topic, _ProducerConfig = []),
+  {ok, PartitionCount} = brod:get_partitions_count(ClientId, Topic),
+  Partitions = lists:seq(0, PartitionCount - 1),
+  %% spawn N + 1 consumers, one of them will have no assignment
+  %% it should work as a 'standing-by' consumer.
+  ok = spawn_consumers(ClientId, Topic, PartitionCount + 1),
+  ok = spawn_producers(ClientId, Topic, DelaySeconds, Partitions),
+  ok.
+
+%% @doc Initialize nothing in our case.
+init(_GroupId, _Arg) -> {ok, []}.
+
+%% @doc Handle one message (not message-set).
+handle_message(_Topic, Partition, Message, State) ->
+  #kafka_message{ offset = Offset
+                , value  = Value
+                } = Message,
+  Seqno = list_to_integer(binary_to_list(Value)),
+  Now = os_time_utc_str(),
+  error_logger:info_msg("~p ~p ~s: offset:~w seqno:~w\n",
+                        [self(), Partition, Now, Offset, Seqno]),
+  {ok, ack, State}.
+
+%%%_* Internal Functions =======================================================
+
+spawn_consumers(ClientId, Topic, ConsumerCount) ->
+  %% commit offsets to kafka every 10 seconds
+  GroupConfig = [{offset_commit_policy, commit_to_kafka_v2}
+                ,{offset_commit_interval_seconds, 5}
+                ],
+  GroupId = iolist_to_binary([Topic, "-group-id"]),
+  lists:foreach(
+    fun(_I) ->
+      {ok, _Subscriber} =
+        brod:start_link_group_subscriber(ClientId, GroupId, [Topic],
+                                         GroupConfig,
+                                         _ConsumerConfig  = [{begin_offset, -2}],
+                                         _CallbackModule  = ?MODULE,
+                                         _CallbackInitArg = [])
+    end, lists:seq(1, ConsumerCount)).
```

Comment:
The main point is why do we need ConsumerCount group subscribers. Shouldn't a single one suffice? It will spawn a consumer for every partition.